linalg/hypre.hpp

```diff
@@ -119,18 +119,18 @@ class HypreParVector : public Vector
    explicit HypreParVector(ParFiniteElementSpace *pfes);
 
    /// MPI communicator
-   MPI_Comm GetComm() { return x->comm; }
+   MPI_Comm GetComm() const { return x->comm; }
 
    /// Converts hypre's format to HypreParVector
    void WrapHypreParVector(hypre_ParVector *y, bool owner=true);
 
    /// Returns the parallel row/column partitioning
    /** See @ref hypre_partitioning_descr "here" for a description of the
        partitioning array. */
-   inline HYPRE_BigInt *Partitioning() { return x->partitioning; }
+   inline HYPRE_BigInt *Partitioning() const { return x->partitioning; }
```

Comment:
If we want this method to be `const` the returned internal should be `const` too I think:
```suggestion
   inline const HYPRE_BigInt *Partitioning() const { return x->partitioning; }
```

Reply:
I agree. Good catch.

Reply:
Adding a const version is fine but why remove the non-const?

Reply:
Is modifying the partitioning array even something that is a well-defined operation after construction of the vector?

Reply:
It's more of a convenience thing -- you may want to get the partition to pass it to some hypre function, avoiding the `const_cast` if your hypre vector is not `const`.

Reply:
If that's something useful that users really do in their code then we can definitely bring back the non-const accessor (even without it you can get the non-const partitioning by casting to `hypre_ParVector*` and accessing the `partitioning` field directly if needed).

Do you know if this is something that exists in the wild somewhere?

Reply:
It is one line and preserves backward compatibility -- I see no reason to remove it.

Reply:
The reason would be to disallow the user from mutating the partitioning of an already created vector (which in my understanding is not a valid thing to do — but I may be wrong about this).

Reply:
For me, this is not a good enough reason to break backward compatibility.

Reply:
Ok, that's fair, and we can add the non-const version (perhaps as `MFEM_DEPRECATED`?) but I still think it's relevant to know if there is any code out there that depends on this, or whether it's purely theoretical.

Reply:
"Convenience thing" is the enemy of good software... We don't want to write a software that's easy to hack in, we want something hard to use incorrectly... I think the `const_cast` is actually good, it pushes a user to think about what he's actually doing. It should feel like a hack to modify directly the partitioning array outside its class. If the partitioning array needs to be modified, then there is most likely a better abstraction for that than giving away the internals.